### PR TITLE
fix!: improve wallet key derivation by use of proper domain separation (see issue  #4170)

### PR DESCRIPTION
**Pull request by jorgeantonio21**
Description
--- Add domain separation for wallet key derivation for type `KeyManager`.

Motivation and Context
--- The current wallet key derivation uses plain concatenation of variable length inputs. This procedure is known to not be collision resistant. A better methodology, as pointed [here](https://github.com/tari-project/tari/issues/4170), is to prepend the length of each datum to it before passing it through the hash function. We attain this by using the current hashing [API](https://github.com/tari-project/tari-crypto/blob/main/src/hashing.rs), which gives a suitable interface for domain separation, (which always prepends input length).

How Has This Been Tested?
--- Unit tests.

### CODE REVIEW

**Comment by delta1** [comment]
Just out of interest this is a breaking change

**Comment by AaronFeickert** [comment]
Length extension attack resistance isn't strictly a requirement here, as key derivation security doesn't depend on it. However, it is the case that any underlying hash function susceptible to length extension is probably a poor choice for other design or performance reasons anyway.

File: base_layer/key_manager/src/key_manager.rs [hunk 1]
```diff
@@ -84,13 +87,17 @@ where
 
     /// Derive a new private key from master key: derived_key=SHA256(master_key||branch_seed||index)
```
Comment (AaronFeickert) [inline base_layer/key_manager/src/key_manager.rs]: This comment is no longer accurate, as `SHA-256` is not resistant to length extension attacks, and the input structure implied by the comment is out of date and no longer correct.

File: base_layer/key_manager/src/key_manager.rs [hunk 2]
```diff
@@ -84,13 +87,17 @@ where
...
-            self.seed.entropy().to_vec().to_hex(),
-            self.branch_seed,
-            key_index
-        );
-        match K::from_bytes(D::digest(&concatenated.into_bytes()).as_slice()) {
+        // apply domain separation to generate derive key. Under the hood, the hashing api prepends the length of each
+        // piece of data for concatenation, reducing the risk of collisions due to redundance of variable length
+        // input
+        let derive_key = DomainSeparatedHasher::<D, GenericHashDomain>::new(DOMAIN_SEPARATION_LABEL)
+            .chain(self.seed.entropy().to_vec().to_hex())
```
Comment (AaronFeickert) [inline base_layer/key_manager/src/key_manager.rs]: Any reason to use a hex encoding here? It seems unnecessary, and the use of the hashing API is already a breaking change.

**Review by AaronFeickert — CHANGES_REQUESTED** [review]

File: base_layer/key_manager/src/key_manager.rs (hunk 1 above)
Comment (jorgeantonio21) [reply base_layer/key_manager/src/key_manager.rs]: Absolutely ! Thank you for taking notice of it !

File: base_layer/key_manager/src/key_manager.rs (hunk 2 above)
Comment (jorgeantonio21) [reply base_layer/key_manager/src/key_manager.rs]: You are totally right Aaron, one can simply chain the seed entropy as it is.

**Review by stringhandler — DISMISSED** [review]

File: base_layer/key_manager/src/lib.rs [hunk 3]
```diff
@@ -11,3 +11,25 @@ pub mod mnemonic_wordlists;
...
+    HashingDomain::new("base_layer.key_manager.cipher_seed.mac_generation")
+}
+
+pub fn base_layer_key_manager_argon2_encoding() -> HashingDomain {
+    HashingDomain::new("base_layer.key_manager.cipher_seed.argon2_encoding")
+}
+
+pub fn base_layer_key_manager_chacha20_encoding() -> HashingDomain {
+    HashingDomain::new("base_layer.key_manager.cipher_seed.chacha20_encoding")
+}
```
Comment (hansieodendaal) [inline base_layer/key_manager/src/lib.rs]: I see these functions are not being used in this PR; I suppose they were created for use elsewhere?

**Review by hansieodendaal — DISMISSED** [review]
LGTM

File: base_layer/key_manager/src/lib.rs (hunk 3 above)
Comment (hansieodendaal) [reply base_layer/key_manager/src/lib.rs]: Ah I see, used in #4296.

File: base_layer/key_manager/src/lib.rs [hunk 4]
```diff
@@ -11,3 +11,25 @@ pub mod mnemonic_wordlists;
...
+
+pub fn base_layer_key_manager() -> HashingDomain {
+    HashingDomain::new("base_layer.key_manager.key_manager")
```
Comment (hansieodendaal) [inline base_layer/key_manager/src/lib.rs]:
```suggestion
    HashingDomain::new("base_layer.key_manager")
```

**Review by stringhandler — APPROVED** [review]

**Pull request merged**